docs/confluence/advanced/grounding.md

```diff
@@ -0,0 +1,154 @@
+
+ 
+### Recursion
+
+```
+assume forall x. f(x) > f(x+1)
+assert f(8) > f(6)
+```
+
+Any sort of such recursion will very likely give a wrong counter-example. We do not auto-detect it, so the responsibility to avoid it and to realize the counter-example is wrong is on the user.
+
+ 
+### Variables That Are Not Arguments
+
+```
+forall x. x^2 != y
+```
+
+Can’t work, because for grounding we run over all instances of functions in the formula. Here there is no function and no instance. This won’t work either:
+```
+forall x, y. f(x) != 2y
+```
+
+Because of `y`. 
+
+We will hard crash on such cases, so again, you can rewrite your spec or turn grounding off. To avoid it, make sure each quantified variable is used in as an argument of a function call within the quantified expression. It’s not always possible, but in this case we can write:
+
+```
+forall x. f(x) % 2 == 1
+```
+ 
+### Double Polarity
+
+The polarity of a sub-formula is the direction of the effect it has on the output of the formula. Best explained by example:
+```
+a /\ (b \/ !c)
+```
+
+`b` has positive polarity, because switching it from `false` to `true`, may only make the formula change from `false` to `true`. In other words, the formula is weakly monotone increasing in `b`. The same goes for `a`. `c` on the other hand has negative polarity.
+
+Sub-expressions can also have double polarity:
+```
+(a == b) /\ c
+ite(a, b, c)
+```
+
+In both of these examples, `a` has double polarity, because the formula is not weakly monotone w.r.t. `a` (neither increasing nor decreasing). 
+
+Grounding crashes when this happens, so either rewrite or turn it off. This is not about the sub-expressions of the quantified expression, but the quantified expression itself. For example:
+```
+ite(forall .., x, y)
+```
+
+is bad. But:
+```
+forall a, b. ite(a, b, b + 1) > 10
+```
+
+is totally fine.
+
+ 
+
+ 
+### Simple Parameters
+
+To make grounding sound, we had to make it work only when the quantified variables appear as the exact argument to some function. For example, 
+```
+forall x. f(2x) > 0
+```
+
+will most likely give a wrong counter-example because `2x` is not a quantified variable. It should be rewritten as:
+```
+forall y. (y % 2 = 0) => f(y) 
+```
+
+Another example:
+```
+forall x. f(x) = g(x+1)
+```
+
+We can’t make both arguments simple at the same time. However, the fact that `x` is simple in one place is many times enough. If it’s not (you got a bad counter-example) you can add another axiom:
+```
+forall x. f(x-1) = g(x)
+```
+
+It’s equivalent, yet for grounding it is different. The first quantifier will be grounded on all `f` instances, and the second on all `g` instances. There is also another solution:
+```
+forall x, y. y=x+1 /\ f(x) = g(y)
+```
+
+This is correct but is slightly less efficient because in general, the fewer variables we quantify over, the smaller the grounding effort will be. Think of it as nested loops.
```

Comment:
This is correct, although slightly less efficient because the fewer the variables we quantify over, the smaller the grounding effort will be. You can think of it as nested loops.